src/widget/textbox.rs

```diff
@@ -150,6 +150,24 @@ impl TextBoxRaw {
         self.selection.end
     }
 
+    /// For a given point, returns the cooresponding offset (in bytes) of
+    /// the grapheme cluster closest to that point.
+    fn offset_for_point(
+        &self,
+        point: Point,
+        piet_text: &mut PietText,
+        data: &String,
+        env: &Env,
+    ) -> usize {
+        let layout = self.get_layout(piet_text, env, data);
+
+        let hscrolled_point = Point::new(point.x + self.hscroll_offset, point.y);
+
+        let hit_test = layout.hit_test_point(hscrolled_point);
+        //TODO: how should we be handling when we're outside the text?
```

Comment:
a question for @hwchen: it wasn't totally clear from the docs in piet how this API is supposed to be used. This might be something to clarify somewhat? In particular, the `is_text` property on `HitTestMetrics` isn't described, and it would be nice if there was a bit more information on what `is_inside` communicates, and how it relates to the returned `text_position`, or how `text_position` is determined for clicks that fall outside the bounds of the text?

Reply:
Sorry for lack of clarity, I should go back and document a little better.

I think my comments here on `is_inside` and `text_position` should be helpful. `is_text` is a bit trickier (I think I messed up), and I will be discussing with @raphlinus how to handle it better. I think that for basic cursor positioning you can safely ignore it.

For `hit_test_point`:
- `is_inside` returns `true` if the point be tested falls within the bounds of the text (basically just the text width).
- some `text_position` will always be returned; if the tested point is inside, it returns the appropriate text position; if it's outside, it will return the nearest text position (either `0` or `text.len()`).

As for `is_text`:
- `is_text` should return true if there's text at that text position.
- You shouldn't have to worry about it for outside-of-text situations. it _should_ return `true` as long as there's text at the `text_position` returned.
- I just dug a little, because of an itch in my head, and it looks like web/cairo can return `false` when the point tested is outside of the text, which is inconsistent with directwrite. This would be a bug in the current implementation.
- `is_text` kind of slipped through from my initial port from directwrite. I'm not sure it's really needed with the current api. An example of usage (to draw selected images): https://github.com/microsoft/Windows-classic-samples/blob/master/Samples/Win7Samples/multimedia/DirectWrite/PadWrite/TextEditor.cpp#L520
- I'll be discussing with @raphlinus soon. Sorry!

Reply:
those are good answers, thanks! Definitely would be nice to get that into the docs at some point, but no sweat. :)